cocos/renderer/scene/camera.js

```diff
@@ -357,59 +364,33 @@ export default class Camera {
    * @returns {vec3} the resulting vector
    */
   screenToWorld(out, screenPos, width, height) {
-    let aspect = width / height;
+    this._calcMatrices(width, height);
+
     let cx = this._rect.x * width;
     let cy = this._rect.y * height;
     let cw = this._rect.w * width;
     let ch = this._rect.h * height;
 
-    // view matrix
-    this._node.getWorldRT(_matView);
-    mat4.invert(_matView, _matView);
-
-    // projection matrix
-    if (this._projection === enums.PROJ_PERSPECTIVE) {
-      mat4.perspective(_matProj,
-        this._fov,
-        aspect,
-        this._near,
-        this._far
-      );
-    } else {
-      let x = this._orthoHeight * aspect;
-      let y = this._orthoHeight;
-      mat4.ortho(_matProj,
-        -x, x, -y, y, this._near, this._far
-      );
-    }
-
-    // view-projection
-    mat4.mul(_matViewProj, _matProj, _matView);
-
-    // inv view-projection
-    mat4.invert(_matInvViewProj, _matViewProj);
-
-    //
     if (this._projection === enums.PROJ_PERSPECTIVE) {
       // calculate screen pos in far clip plane
       vec3.set(out,
-        (screenPos.x - cx) * 2.0 / cw - 1.0,
-        (screenPos.y - cy) * 2.0 / ch - 1.0, // DISABLE: (ch - (screenPos.y - cy)) * 2.0 / ch - 1.0,
+        (screenPos.x - cx) / cw * 2 - 1,
+        (screenPos.y - cy) / ch * 2 - 1,
         1.0
       );
 
       // transform to world
       vec3.transformMat4(out, out, _matInvViewProj);
 
-      //
+      // lerp to depth z
       this._node.getWorldPosition(_tmp_v3);
-      vec3.lerp(out, _tmp_v3, out, screenPos.z / this._far);
+
+      vec3.lerp(out, _tmp_v3, out, lerp(this._near, this._far, screenPos.z));
```

Comment:
seems should be ` vec3.lerp(out, _tmp_v3, out, lerp(0, this._far, screenPos.z));`, coz `_tmp_v3` is the position of camera not a point on the near plane.